### PR TITLE
backlog complete

### PR DESCRIPTION
**Pull request by rayzhang-xingyunzh**

### CODE REVIEW

File: controllers/userStoryController.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+		console.log(err);
+		res.send(util.wrapBody('Internal Error','E'));
+	});
+};
+
+function unionObjectIds(a,b){
+
+	for (var i = b.length - 1; i >= 0; i--) {
+		if (!b[i]) {
+			delete b[i];
```
Comment (williamyu-xingyunzh) [inline controllers/userStoryController.js:69]: This is dangerous. You should not delete elements when iterating array.

File: controllers/userStoryController.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+			continue;
+		}
+
+		var flag = true;
+
+		for (var j = a.length - 1; j >= 0; j--) {
+			if (!a[j]) continue;
+			if (b[i].toString() == a[j].toString()) {
```
Comment (williamyu-xingyunzh) [inline controllers/userStoryController.js:77]: toString is not a good way to compare obj.

File: controllers/userStoryController.js [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+		}
+
+		if (flag) {
+			a.push(b[i]);
+		}
+
+
+	}
+
+	return a;
```
Comment (williamyu-xingyunzh) [inline controllers/userStoryController.js:89]: I recommend you changing the algorithm. You can use Set to do it in a N*lgN time, and more simple code.

File: controllers/userStoryController.js [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+function updateUserStories(usChanges,count){
+	if (count === undefined) {
+		count = usChanges.length;
+	}else if(count === 0){
+		return usChanges;
+	}
+
+	count--;
```
Comment (williamyu-xingyunzh) [inline controllers/userStoryController.js:99]: generally, it is not a good idea to have such operation here.  Difficult to read and get the meaning.

File: controllers/userStoryController.js [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+			//update
+			promise = userStoryRepository.updateById(usChange._id,usChange);	
+		}
+	}else{
+		//create
+		promise = userStoryRepository.create(usChange);
+	}
+
+	return promise.then(function(usChange){
+		usChanges[count] = usChange;
```
Comment (williamyu-xingyunzh) [inline controllers/userStoryController.js:117]: If usChange got removed like line 106, would it make sense to assign it back to array? Also it's not easy to understand why you do it in a recursive way.  Is a for loop enough for this secenario?

File: repositories/projectRepository.js [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+	})
+	.populate({
+		path:'relatedIdea manager backlog',
+		populate:{
+			path:'innovator consultant'
+		}
+	}).lean().exec();
+};
+
+exports.create = function(data){
```
Comment (williamyu-xingyunzh) [inline repositories/projectRepository.js:68]: Is this method needed?

File: repositories/projectRepository.js [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+ #param options
+	pageNum
+	pageSize
+	sector
+	keywords
+*/
+exports.query = function(options){
+	
+
+	var conditions = {};
```
Comment (williamyu-xingyunzh) [inline repositories/projectRepository.js:82]: I can see only name is injected to conditions, how about other parameters?

File: repositories/taskRepository.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
+var Task = require('../models/task');
+
+exports.getTaskById = function(id) {
+	return Task.findById(id).lean().exec();
```
Comment (williamyu-xingyunzh) [inline repositories/taskRepository.js:4]: why does it not have populate step?

File: routes/ideaAPI.js [hunk 9]
```diff
@@ -16,12 +16,16 @@ var router = require('express').Router();
...
+//{idea:IdeaEntity}
 router.post('/add',authenticator.authenticate,ideaController.createIdea);
 
 //query parameters
 //	required:
 //		id:String (idea)
-router.get('/id/:id',ideaController.getIdeaById);
+//response:
+//{idea:IdeaEntity}
+router.get('/:id',ideaController.getIdeaById);
```
Comment (williamyu-xingyunzh) [inline routes/ideaAPI.js:28]: why do you change back? Interface should be separated well.  You mixed the "/actiond_name" and "/:parameter" pattern.

File: routes/router.js [hunk 10]
```diff
@@ -29,4 +32,8 @@ module.exports = function(app, contextRoot) {
 
 	rootRouter.use('/api/team',teamAPI);
 
+	rootRouter.use('/api/project',projectAPI);
+
+    rootRouter.use('/api/us',userStoryAPI);
```
Comment (williamyu-xingyunzh) [inline routes/router.js:37]: us is not a good name, it is well known as a country name. better to use "story".

File: routes/sprintAPI.js [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+var authenticator = require('../authenticate/authenticator');
+var router = require('express').Router();
+
+router.get('/project/:id',sprintController.getSprintsByProject);
+
+router.post('/update/:id',sprintController.update);
+
+router.post('/add',sprintController.create);
+
+router.get('/:id',sprintController.getSprintById);
```
Comment (williamyu-xingyunzh) [inline routes/sprintAPI.js:11]: some as idea part. Mixing the interface pattern is not a good design.

File: routes/teamAPI.js [hunk 12]
```diff
@@ -11,26 +11,34 @@ var router = require('express').Router();
...
 router.post('/list',teamController.list);
 
 
-//query parameters
+//body parameters
 //	required:
 //		name:String 
+//response:
+//{exist:true/false}
 router.post('/check/name',teamController.checkNameExist);
```
Comment (williamyu-xingyunzh) [inline routes/teamAPI.js:42]: I remember I had this comment before, why do you need the second slash? Also why making it a post method. In this case, get method is more fit. It is a purely get action.   Please adhere to http/REST standards recommendation as much as possible.

File: routes/teamAPI.js [hunk 13]
```diff
@@ -11,26 +11,34 @@ var router = require('express').Router();
...
 //body parameters
 //	required:
 //	optional:
-//		pageNum:Number (default:1)
+//		pageNum:Number (default:0)
 //		pageSize:Number (default:10)
 //		keyword:String (for name)
+//response:
+// {total:totalNumber,teams:[TeamEntities]}
 router.post('/list',teamController.list);
```
Comment (williamyu-xingyunzh) [inline routes/teamAPI.js:34]: get method is prefered in this case.

File: routes/userStoryAPI.js [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+//response:
+// {backlog:[UserStoryEntities]}
+router.get('/project/:id',userStoryController.getUSByProject);
+
+//query parameters
+//	required:
+//		id:String (userStory)
+//response:
+// {userStory:{UserStoryEntity}
+router.get('/:id',userStoryController.getUserStoryById);
```
Comment (williamyu-xingyunzh) [inline routes/userStoryAPI.js:17]: Mixed pattern. Same as above

**Review by williamyu-xingyunzh — APPROVED** [review]
Please read the comments and discuss with me if needed. Some of them must be fixed in next run.

File: routes/userStoryAPI.js (hunk 14 above)
Comment (rayzhang-xingyunzh) [reply routes/userStoryAPI.js:17]: fixed

File: repositories/projectRepository.js (hunk 6 above)
Comment (rayzhang-xingyunzh) [inline repositories/projectRepository.js]: not so far.fixed

File: repositories/projectRepository.js (hunk 7 above)
Comment (rayzhang-xingyunzh) [reply repositories/projectRepository.js:82]: fixed

File: routes/teamAPI.js (hunk 12 above)
Comment (rayzhang-xingyunzh) [reply routes/teamAPI.js:42]: fixed

File: routes/sprintAPI.js (hunk 11 above)
Comment (rayzhang-xingyunzh) [reply routes/sprintAPI.js:11]: fixed

File: routes/router.js (hunk 10 above)
Comment (rayzhang-xingyunzh) [reply routes/router.js:37]: fixed

File: routes/ideaAPI.js (hunk 9 above)
Comment (rayzhang-xingyunzh) [reply routes/ideaAPI.js:28]: fixed

**Pull request merged**